File: Backend/pdf_service.py
"""
PDF Service for server-side processing (Web requests only)
"""
import io
from typing import List, Dict, Any
from datetime import datetime
from PyPDF2 import PdfReader, PdfWriter
from reportlab.lib.pagesizes import A4, A3, LETTER
from reportlab.lib.units import mm
from reportlab.pdfgen import canvas
from reportlab.lib import colors

# Paper size mapping
PAPER_SIZES = {
    'A4': A4,
    'A3': A3,
    'LETTER': LETTER,
}

def create_cover_sheet(metadata: Dict[str, Any]) -> bytes:
    """
    Generate a cover sheet PDF with order details.
    Only used for web requests (processing_mode == 'server')
    """
    buffer = io.BytesIO()
    
    # Get paper size from config
    paper_size = metadata.get('config', {}).get('paperSize', 'A4')
    page_size = PAPER_SIZES.get(paper_size, A4)
    
    c = canvas.Canvas(buffer, pagesize=page_size)
    width, height = page_size
    
    # Header
    c.setFillColor(colors.HexColor('#6200EE'))
    c.rect(0, height - 80, width, 80, fill=True, stroke=False)
    
    c.setFillColor(colors.white)
    c.setFont('Helvetica-Bold', 24)
    c.drawCentredString(width / 2, height - 50, 'PRINT ORDER')
    
    # Order ID
    order_id = metadata.get('orderId', 'UNKNOWN')
    c.setFont('Helvetica', 14)
    c.drawCentredString(width / 2, height - 70, f'Order ID: {order_id}')
    
    # Student Details Section
    y_pos = height - 120
    c.setFillColor(colors.black)
    c.setFont('Helvetica-Bold', 16)
    c.drawString(40, y_pos, 'Student Details')
    
    y_pos -= 30
    c.setFont('Helvetica', 12)
    student = metadata.get('student', {})
    details = [
        ('Name', student.get('name', 'N/A')),
        ('Student ID', student.get('studentId', 'N/A')),
        ('Phone', student.get('phone', 'N/A')),
    ]
    
    for label, value in details:
        c.drawString(50, y_pos, f'{label}:')
        c.drawString(150, y_pos, str(value))
        y_pos -= 20
    
    # Print Configuration Section
    y_pos -= 20
    c.setFont('Helvetica-Bold', 16)
    c.drawString(40, y_pos, 'Print Configuration')
    
    y_pos -= 30
    c.setFont('Helvetica', 12)
    config = metadata.get('config', {})
    config_details = [
        ('Paper Size', config.get('paperSize', 'A4')),
        ('Print Type', 'Color' if config.get('printType') == 'COLOR' else 'Black & White'),
        ('Print Side', 'Double-sided' if config.get('printSide') == 'DOUBLE' else 'Single-sided'),
        ('Copies', config.get('copies', 1)),
        ('Total Pages', config.get('totalPages', 0)),
    ]
    
    for label, value in config_details:
        c.drawString(50, y_pos, f'{label}:')
        c.drawString(150, y_pos, str(value))
        y_pos -= 20
    
    # Payment Section
    y_pos -= 20
    c.setFont('Helvetica-Bold', 16)
    c.drawString(40, y_pos, 'Payment Details')
    
    y_pos -= 30
    c.setFont('Helvetica', 12)
    payment = metadata.get('payment', {})
    total_price = config.get('totalPrice', 0)
    c.drawString(50, y_pos, f'Total Amount:')
    c.setFont('Helvetica-Bold', 14)
    c.drawString(150, y_pos, f'₹{total_price:.2f}')
    
    y_pos -= 20
    c.setFont('Helvetica', 12)
    if payment.get('transactionId'):
        c.drawString(50, y_pos, f'Transaction ID:')
        c.drawString(150, y_pos, payment.get('transactionId', 'N/A'))
    
    # Timestamp
    y_pos -= 40
    c.setFont('Helvetica', 10)
    c.setFillColor(colors.gray)
    timestamp = datetime.now().strftime('%Y-%m-%d %H:%M:%S')
    c.drawString(50, y_pos, f'Generated: {timestamp}')
    
    # Footer
    c.setFillColor(colors.HexColor('#6200EE'))
    c.rect(0, 0, width, 30, fill=True, stroke=False)
    c.setFillColor(colors.white)
    c.setFont('Helvetica', 10)
    c.drawCentredString(width / 2, 10, 'ABDUL MANNAN MOLLA - Print Service')
    
    c.save()
    buffer.seek(0)
    return buffer.read()


def merge_pdfs(pdf_files: List[bytes], cover_sheet: bytes = None) -> bytes:
    """
    Merge multiple PDF files into one.
    Optionally prepend a cover sheet.
    Only used for web requests (processing_mode == 'server')
    """
    writer = PdfWriter()
    
    # Add cover sheet first if provided
    if cover_sheet:
        cover_reader = PdfReader(io.BytesIO(cover_sheet))
        for page in cover_reader.pages:
            writer.add_page(page)
    
    # Add all document pages
    for pdf_bytes in pdf_files:
        try:
            reader = PdfReader(io.BytesIO(pdf_bytes))
            for page in reader.pages:
                writer.add_page(page)
        except Exception as e:
            # print(f"Error reading PDF: {e}")
            # Skip corrupted PDFs
            continue
    
    # Write merged PDF to buffer
    output = io.BytesIO()
    writer.write(output)
    output.seek(0)
    return output.read()


def get_pdf_page_count(pdf_bytes: bytes) -> int:
    """Get the number of pages in a PDF."""
    try:
        reader = PdfReader(io.BytesIO(pdf_bytes))
        return len(reader.pages)
    except:
        return 0

File: student/lib/services/pdf_service.dart
import 'dart:typed_data';
import 'dart:ui' as ui;
import 'package:flutter/foundation.dart';
import 'package:syncfusion_flutter_pdf/pdf.dart' as syncfusion;
import 'package:pdf/pdf.dart';
import 'package:pdf/widgets.dart' as pw;
import 'package:qr_flutter/qr_flutter.dart';
import '../models/print_order.dart';
import '../utils/helpers.dart';
import '../utils/platform_utils.dart';

/// Service for PDF operations (page count, merging, front page generation)
class PdfService {
  /// Get page count from PDF bytes
  static Future<int> getPageCount(Uint8List pdfBytes) async {
    try {
      final document = syncfusion.PdfDocument(inputBytes: pdfBytes);
      final pageCount = document.pages.count;
      document.dispose();
      return pageCount;
    } catch (e) {
      if (kDebugMode) debugPrint('Error getting page count: $e');
      return 0;
    }
  }

  /// Validate if the bytes represent a valid PDF
  static bool isValidPdf(Uint8List bytes) {
    if (bytes.length < 4) return false;
    // Check PDF magic number
    return bytes[0] == 0x25 && // %
        bytes[1] == 0x50 && // P
        bytes[2] == 0x44 && // D
        bytes[3] == 0x46; // F
  }

  /// Generate front page PDF with order details and QR code
  /// Uses B&W design and matches document page configuration
  static Future<Uint8List> generateFrontPage(PrintOrder order, {Function(String)? onLog}) async {
    void log(String msg) {
      if (kDebugMode) debugPrint('PdfService: $msg');
      onLog?.call('PDF: $msg');
    }
    
    log('Starting front page generation for order ${order.orderId}');
    final pdf = pw.Document();
    
    // SKIP QR code generation - it hangs on Android
    // Just use text-based order ID display instead
    log('Skipping QR code (causes hang), using text display');
    final pw.ImageProvider? qrImage = null;
    
    log('Determining page format...');
    // Determine page format based on document config
    PdfPageFormat pageFormat;
    switch (order.config.paperSize.toUpperCase()) {
      case 'LETTER':
        pageFormat = PdfPageFormat.letter;
        break;
      case 'LEGAL':
        pageFormat = PdfPageFormat.legal;
        break;
      case 'A4':
      default:
        pageFormat = PdfPageFormat.a4;
        break;
    }
    
    log('Adding page to PDF document...');
    pdf.addPage(
      pw.Page(
        pageFormat: pageFormat,
        margin: const pw.EdgeInsets.all(40),
        build: (context) {
          return pw.Column(
            crossAxisAlignment: pw.CrossAxisAlignment.start,
            children: [
              // Header - Simple text only (ink-saving)
              pw.Center(
                child: pw.Column(
                  children: [
                    pw.Text(
                      'PRINT ORDER',
                      style: pw.TextStyle(
                        fontSize: 28,
                        fontWeight: pw.FontWeight.bold,
                        color: PdfColors.black,
                      ),
                    ),
                    pw.SizedBox(height: 4),
                    pw.Text(
                      'College Xerox Shop',
                      style: pw.TextStyle(
                        fontSize: 14,
                        color: PdfColors.grey700,
                      ),
                    ),
                  ],
                ),
              ),
              
              pw.SizedBox(height: 30),
              
              // Order ID and QR Code
              pw.Row(
                mainAxisAlignment: pw.MainAxisAlignment.spaceBetween,
                children: [
                  pw.Column(
                    crossAxisAlignment: pw.CrossAxisAlignment.start,
                    children: [
                      pw.Text(
                        'Order ID',
                        style: pw.TextStyle(
                          fontSize: 12,
                          color: PdfColors.grey700,
                        ),
                      ),
                      pw.SizedBox(height: 4),
                      pw.Text(
                        order.orderId,
                        style: pw.TextStyle(
                          fontSize: 20,
                          fontWeight: pw.FontWeight.bold,
                          color: PdfColors.black,
                        ),
                      ),
                      pw.SizedBox(height: 16),
                      pw.Text(
                        'Date & Time',
                        style: pw.TextStyle(
                          fontSize: 12,
                          color: PdfColors.grey700,
                        ),
                      ),
                      pw.SizedBox(height: 4),
                      pw.Text(
                        formatDate(order.createdAt),
                        style: pw.TextStyle(fontSize: 14, color: PdfColors.black),
                      ),
                    ],
                  ),
                  if (qrImage != null)
                    pw.Container(
                      width: 100,
                      height: 100,
                      child: pw.Image(qrImage),
                    ),
              ],
              ),
              
              pw.SizedBox(height: 30),
              
              // Divider
              pw.Divider(color: PdfColors.grey400),
              
              pw.SizedBox(height: 20),
              
              // Student Details Section (No email)
              _buildSectionBW('Student Details', [
                _buildDetailRowBW('Name', order.student.name),
                _buildDetailRowBW('Enrollment No.', order.student.studentId),
                _buildDetailRowBW('Phone', order.student.phone),
              ]),
              
              pw.SizedBox(height: 24),
              
              // Print Configuration Section
              _buildSectionBW('Print Configuration', [
                _buildDetailRowBW('Paper Size', order.config.paperSize),
                _buildDetailRowBW('Print Type', order.config.printType == 'BW' ? 'Black & White' : 'Color'),
                _buildDetailRowBW('Sides', order.config.printSide == 'SINGLE' ? 'Single Sided' : 'Double Sided'),
                _buildDetailRowBW('Copies', order.config.copies.toString()),
              ]),
              
              pw.SizedBox(height: 24),
              
              // Documents Section
              _buildSectionBW('Documents', [
                for (final file in order.files)
                  _buildDetailRowBW(file.name, '${file.pageCount} pages'),
              ]),
              
              pw.SizedBox(height: 30),
              
              // Summary Box - B&W
              pw.Container(
                width: double.infinity,
                padding: const pw.EdgeInsets.all(16),
                decoration: pw.BoxDecoration(
                  color: PdfColors.grey200,
                  borderRadius: pw.BorderRadius.circular(8),
                  border: pw.Border.all(color: PdfColors.grey500),
                ),
                child: pw.Row(
                  mainAxisAlignment: pw.MainAxisAlignment.spaceBetween,
                  children: [
                    pw.Column(
                      crossAxisAlignment: pw.CrossAxisAlignment.start,
                      children: [
                        pw.Text(
                          'Total Pages: ${order.totalPages}',
                          style: pw.TextStyle(fontSize: 14, color: PdfColors.black),
                        ),
                        pw.Text(
                          'Files: ${order.files.length}',
                          style: pw.TextStyle(fontSize: 14, color: PdfColors.black),
                        ),
                      ],
                    ),
                    pw.Column(
                      crossAxisAlignment: pw.CrossAxisAlignment.end,
                      children: [
                        pw.Text(
                          'Total Amount',
                          style: pw.TextStyle(
                            fontSize: 12,
                            color: PdfColors.grey700,
                          ),
                        ),
                        pw.Text(
                          formatCurrency(order.totalPrice),
                          style: pw.TextStyle(
                            fontSize: 24,
                            fontWeight: pw.FontWeight.bold,
                            color: PdfColors.black,
                          ),
                        ),
                      ],
                    ),
                  ],
                ),
              ),
              
              pw.Spacer(),
              
              // Footer - B&W
              pw.Container(
                width: double.infinity,
                padding: const pw.EdgeInsets.symmetric(vertical: 12),
                decoration: const pw.BoxDecoration(
                  border: pw.Border(
                    top: pw.BorderSide(color: PdfColors.grey400),
                  ),
                ),
                child: pw.Column(
                  children: [
                    pw.Text(
                      'Payment ${order.payment?.isVerified == true ? 'VERIFIED' : 'PENDING'}',
                      style: pw.TextStyle(
                        fontSize: 12,
                        fontWeight: pw.FontWeight.bold,
                        color: PdfColors.black,
                      ),
                    ),
                    if (order.payment?.transactionId != null)
                      pw.Text(
                        'Transaction ID: ${order.payment!.transactionId}',
                        style: pw.TextStyle(
                          fontSize: 10,
                          color: PdfColors.grey700,
                        ),
                      ),
                    pw.SizedBox(height: 8),
                    pw.Text(
                      'Generated on ${PlatformUtils.platformName}',
                      style: pw.TextStyle(
                        fontSize: 9,
                        color: PdfColors.grey600,
                      ),
                    ),
                  ],
                ),
              ),
            ],
          );
        },
      ),
    );
    
    log('Page added. Saving PDF document...');
    final bytes = await pdf.save();
    log('Front page PDF saved successfully: ${bytes.length} bytes');
    return bytes;
  }

  /// Helper to build section widget (B&W version)
  static pw.Widget _buildSectionBW(String title, List<pw.Widget> rows) {
    return pw.Column(
      crossAxisAlignment: pw.CrossAxisAlignment.start,
      children: [
        pw.Text(
          title,
          style: pw.TextStyle(
            fontSize: 16,
            fontWeight: pw.FontWeight.bold,
            color: PdfColors.black,
          ),
        ),
        pw.SizedBox(height: 12),
        ...rows,
      ],
    );
  }

  /// Helper to build detail row (B&W version)
  static pw.Widget _buildDetailRowBW(String label, String value) {
    return pw.Padding(
      padding: const pw.EdgeInsets.only(bottom: 8),
      child: pw.Row(
        crossAxisAlignment: pw.CrossAxisAlignment.start,
        children: [
          pw.SizedBox(
            width: 120,
            child: pw.Text(
              label,
              style: pw.TextStyle(
                fontSize: 12,
                color: PdfColors.grey700,
              ),
            ),
          ),
          pw.Expanded(
            child: pw.Text(
              value,
              style: pw.TextStyle(
                fontSize: 12,
                fontWeight: pw.FontWeight.bold,
                color: PdfColors.black,
              ),
            ),
          ),
        ],
      ),
    );
  }

  /// Helper to build section widget
  static pw.Widget _buildSection(String title, List<pw.Widget> rows) {
    return pw.Column(
      crossAxisAlignment: pw.CrossAxisAlignment.start,
      children: [
        pw.Text(
          title,
          style: pw.TextStyle(
            fontSize: 16,
            fontWeight: pw.FontWeight.bold,
          ),
        ),
        pw.SizedBox(height: 12),
        ...rows,
      ],
    );
  }

  /// Helper to build detail row
  static pw.Widget _buildDetailRow(String label, String value) {
    return pw.Padding(
      padding: const pw.EdgeInsets.symmetric(vertical: 4),
      child: pw.Row(
        crossAxisAlignment: pw.CrossAxisAlignment.start,
        children: [
          pw.SizedBox(
            width: 120,
            child: pw.Text(
              label,
              style: pw.TextStyle(
                fontSize: 12,
                color: PdfColors.grey600,
              ),
            ),
          ),
          pw.Expanded(
            child: pw.Text(
              value,
              style: const pw.TextStyle(fontSize: 12),
            ),
          ),
        ],
      ),
    );
  }

  /// Generate QR code image for PDF
  static Future<pw.ImageProvider?> _generateQrImage(String data) async {
    try {
      if (kDebugMode) debugPrint('PdfService: Creating QR painter...');
      final qrPainter = QrPainter(
        data: data,
        version: QrVersions.auto,
        errorCorrectionLevel: QrErrorCorrectLevel.M,
      );
      
      if (kDebugMode) debugPrint('PdfService: Converting QR to image...');
      final image = await qrPainter.toImage(200).timeout(
        const Duration(seconds: 10),
        onTimeout: () {
          if (kDebugMode) debugPrint('PdfService: QR toImage TIMEOUT!');
          throw Exception('QR code generation timed out');
        },
      );
      
      if (kDebugMode) debugPrint('PdfService: Getting byte data from QR image...');
      final byteData = await image.toByteData(format: ui.ImageByteFormat.png).timeout(
        const Duration(seconds: 10),
        onTimeout: () {
          if (kDebugMode) debugPrint('PdfService: toByteData TIMEOUT!');
          return null;
        },
      );
      if (byteData == null) {
        if (kDebugMode) debugPrint('PdfService: QR byteData is null');
        return null;
      }
      
      if (kDebugMode) debugPrint('PdfService: QR image generated successfully (${byteData.lengthInBytes} bytes)');
      return pw.MemoryImage(byteData.buffer.asUint8List());
    } catch (e) {
      if (kDebugMode) debugPrint('Error generating QR image: $e');
      return null;
    }
  }

  /// Merge multiple PDFs into one while preserving original page sizes
  static Future<Uint8List> mergePdfs(List<Uint8List> pdfBytesList, {Function(String)? onLog}) async {
    void log(String msg) {
      if (kDebugMode) debugPrint('PdfService: $msg');
      onLog?.call('PDF: $msg');
    }
    
    log('Starting PDF merge of ${pdfBytesList.length} documents');
    final mergedDocument = syncfusion.PdfDocument();
    int docIndex = 0;
    
    for (final pdfBytes in pdfBytesList) {
      log('Processing document $docIndex (${pdfBytes.length} bytes)...');
      docIndex++;
      try {
        log('  Loading document...');
        final document = syncfusion.PdfDocument(inputBytes: pdfBytes);
        log('  Document has ${document.pages.count} pages');
        
        for (int i = 0; i < document.pages.count; i++) {
          log('  Processing page ${i + 1}/${document.pages.count}...');
          final sourcePage = document.pages[i];
          final template = sourcePage.createTemplate();
          
          // Add a new section for this page to control its size independent of others
          final section = mergedDocument.sections!.add();
          
          // Set the page size to match source (use size, not clientSize to include margins)
          section.pageSettings.size = sourcePage.size;
          section.pageSettings.margins.all = 0;
          
          // Add page to this section
          final page = section.pages.add();
          
          // Draw the template
          page.graphics.drawPdfTemplate(
            template,
            ui.Offset.zero,
            sourcePage.size,
          );
        }
        document.dispose();
        log('  Document $docIndex processed successfully');
      } catch (e) {
        log('ERROR merging PDF document $docIndex: $e');
      }
    }
    
    log('Saving merged PDF...');
    final mergedBytes = Uint8List.fromList(mergedDocument.saveSync());
    mergedDocument.dispose();
    log('Merged PDF saved: ${mergedBytes.length} bytes');
    
    return mergedBytes;
  }

  /// Merge front page with document PDFs
  static Future<Uint8List> mergeWithFrontPage(
    Uint8List frontPageBytes, 
    List<Uint8List> documentBytesList,
    {Function(String)? onLog}
  ) async {
    onLog?.call('PDF: Merging front page with ${documentBytesList.length} documents...');
    return mergePdfs([frontPageBytes, ...documentBytesList], onLog: onLog);
  }
}
